File: apps/dsa-web/src/api/stocks.ts
import apiClient from './index';
import type { KLine, ResonanceLevel } from '../types/kline';
import type { SignalMarker, SignalsResponse } from '../types/kline';
import type { BoardEntry, SignalsBoardResponse } from '../types/kline';
import type { StockQuote } from '../types/kline';

export type ExtractItem = {
  code?: string | null;
  name?: string | null;
  confidence: string;
};

export type ExtractFromImageResponse = {
  codes: string[];
  items?: ExtractItem[];
  rawText?: string;
};

/** K 线抽屉默认回看天数，与后端 /history 端点 Query 默认对齐。 */
export const KLINE_DEFAULT_DAYS = 120;

type RawSignalMarker = {
  timestamp: number;
  price: number;
  anchor: SignalMarker['anchor'];
  direction: SignalMarker['direction'];
  signal_type: string;
  source: SignalMarker['source'];
  confidence: SignalMarker['confidence'];
  is_daily_approx: boolean;
  is_anomalous: boolean;
  reason: string;
  threshold: number | null;
  observed_value: number | null;
  hit_rate: number | null;
  hit_sample: number | null;
  verified: boolean;
  ci_low: number | null;
  ci_high: number | null;
  baseline_excess: number | null;
  ci_low_corrected?: number | null;
  family_size?: number | null;
  as_of: number | null;
  horizon_bars?: number | null;
  status?: 'active' | 'aging' | 'expired' | null;
};

type RawSignalsResponse = {
  status: SignalsResponse['status'];
  consistency: SignalsResponse['consistency'];
  degraded_reason: string | null;
  price_lines: { entry: number | null; stop: number | null; target: number | null } | null;
  markers: RawSignalMarker[] | null;
  resonance?: ResonanceLevel | null;
  plan_quality?: 'high' | 'medium' | 'low' | null;
};

const mapSignalMarker = (raw: RawSignalMarker): SignalMarker => ({
  timestamp: raw.timestamp,
  price: raw.price,
  anchor: raw.anchor,
  direction: raw.direction,
  signalType: raw.signal_type,
  source: raw.source,
  confidence: raw.confidence,
  isDailyApprox: raw.is_daily_approx,
  isAnomalous: raw.is_anomalous,
  reason: raw.reason,
  threshold: raw.threshold ?? null,
  observedValue: raw.observed_value ?? null,
  hitRate: raw.hit_rate ?? null,
  hitSample: raw.hit_sample ?? null,
  verified: raw.verified,
  ciLow: raw.ci_low ?? null,
  ciHigh: raw.ci_high ?? null,
  baselineExcess: raw.baseline_excess ?? null,
  ciLowCorrected: raw.ci_low_corrected ?? null,
  familySize: raw.family_size ?? null,
  asOf: raw.as_of ?? null,
  horizonBars: raw.horizon_bars ?? null,
  status: raw.status ?? null,
});

type RawBoardEntry = {
  code: string; name: string | null; market: string | null;
  action_group: BoardEntry['actionGroup'];
  rule_direction: BoardEntry['ruleDirection']; llm_direction: BoardEntry['llmDirection'];
  consistency: BoardEntry['consistency']; key_signals: string[];
  price_lines: { entry: number | null; stop: number | null; target: number | null };
  latest_close: number | null; hit_rate: number | null; hit_sample: number | null;
  verified: boolean; ci_low: number | null; ci_high: number | null; baseline_excess: number | null;
  ci_low_corrected?: number | null;
  family_size?: number | null;
  status: BoardEntry['status']; degraded_reason: string | null;
  resonance?: ResonanceLevel | null;
  horizon_bars?: number | null;
  signal_status?: 'active' | 'aging' | 'expired' | null;
  plan_quality?: 'high' | 'medium' | 'low' | null;
};
type RawBoardResponse = {
  as_of: number; entries: RawBoardEntry[] | null;
  counts: { buy: number; hold: number; sell: number; unavailable: number };
  degraded_codes: string[] | null;
};
const mapBoardEntry = (r: RawBoardEntry): BoardEntry => ({
  code: r.code, name: r.name ?? null, market: r.market ?? null,
  actionGroup: r.action_group, ruleDirection: r.rule_direction ?? null,
  llmDirection: r.llm_direction ?? null, consistency: r.consistency,
  keySignals: r.key_signals ?? [],
  priceLines: { entry: r.price_lines?.entry ?? null, stop: r.price_lines?.stop ?? null, target: r.price_lines?.target ?? null },
  latestClose: r.latest_close ?? null, hitRate: r.hit_rate ?? null, hitSample: r.hit_sample ?? null,
  verified: r.verified, ciLow: r.ci_low ?? null, ciHigh: r.ci_high ?? null, baselineExcess: r.baseline_excess ?? null,
  ciLowCorrected: r.ci_low_corrected ?? null,
  familySize: r.family_size ?? null,
  status: r.status, degradedReason: r.degraded_reason ?? null,
  resonance: r.resonance ?? 'none',
  horizonBars: r.horizon_bars ?? null,
  signalStatus: r.signal_status ?? null,
  planQuality: r.plan_quality ?? null,
});

export const stocksApi = {
  async extractFromImage(file: File): Promise<ExtractFromImageResponse> {
    const formData = new FormData();
    formData.append('file', file);

    const headers: { [key: string]: string | undefined } = { 'Content-Type': undefined };
    const response = await apiClient.post(
      '/api/v1/stocks/extract-from-image',
      formData,
      {
        headers,
        timeout: 60000, // Vision API can be slow; 60s
      },
    );

    const data = response.data as { codes?: string[]; items?: ExtractItem[]; raw_text?: string };
    return {
      codes: data.codes ?? [],
      items: data.items,
      rawText: data.raw_text,
    };
  },

  async parseImport(file?: File, text?: string): Promise<ExtractFromImageResponse> {
    if (file) {
      const formData = new FormData();
      formData.append('file', file);
      const headers: { [key: string]: string | undefined } = { 'Content-Type': undefined };
      const response = await apiClient.post('/api/v1/stocks/parse-import', formData, { headers });
      const data = response.data as { codes?: string[]; items?: ExtractItem[] };
      return { codes: data.codes ?? [], items: data.items };
    }
    if (text) {
      const response = await apiClient.post('/api/v1/stocks/parse-import', { text });
      const data = response.data as { codes?: string[]; items?: ExtractItem[] };
      return { codes: data.codes ?? [], items: data.items };
    }
    throw new Error('请提供文件或粘贴文本');
  },

  /**
   * 拉取日线 K 线，映射为 klinecharts 所需的 KLine[]。
   * 与 /signals 同源同 days（M0 暂只用 /history）；不复用 api/history.ts（分析记录域）。
   * code 经 encodeURIComponent 以兼容带 '/' 的 crypto 代码（后端 {code:path} 路由）。
   */
  async getKlineHistory(
    code: string,
    days: number = KLINE_DEFAULT_DAYS,
    period: 'daily' | 'weekly' | 'monthly' = 'daily',
  ): Promise<KLine[]> {
    const response = await apiClient.get(
      `/api/v1/stocks/${encodeURIComponent(code)}/history`,
      { params: { days, period } },
    );
    const data = response.data as { data?: KLineDataRaw[] };
    const rows = data.data ?? [];
    return rows
      .map(mapKLineDataToKLine)
      .sort((a, b) => a.timestamp - b.timestamp);
  },

  /**
   * 拉取量价信号（规则 + LLM 双轨），映射 snake_case → camelCase。
   * 与 getKlineHistory 同源同 days；degraded/空数据保持后端形状，不静默吞错。
   */
  async getSignals(code: string, days?: number): Promise<SignalsResponse> {
    const params: { days?: number } = {};
    if (days !== undefined) {
      params.days = days;
    }
    // code 经 encodeURIComponent 兼容带 '/' 的 crypto 代码（后端 {code:path} 路由），与 getKlineHistory 同源。
    const response = await apiClient.get(
      `/api/v1/stocks/${encodeURIComponent(code)}/signals`,
      { params },
    );
    const data = response.data as RawSignalsResponse;
    return {
      status: data.status,
      consistency: data.consistency,
      degradedReason: data.degraded_reason ?? null,
      priceLines: {
        entry: data.price_lines?.entry ?? null,
        stop: data.price_lines?.stop ?? null,
        target: data.price_lines?.target ?? null,
      },
      markers: (data.markers ?? []).map(mapSignalMarker),
      resonance: data.resonance ?? 'none',
      planQuality: data.plan_quality ?? null,
    };
  },

  /**
   * 拉取自选股信号看板（多股聚合），映射 snake_case → camelCase。
   * days 透传至单股信号窗口；refresh=true 时跳过后端 TTL 缓存。
   */
  async getBoard(days?: number, refresh?: boolean): Promise<SignalsBoardResponse> {
    const params: { days?: number; refresh?: boolean } = {};
    if (days !== undefined) params.days = days;
    if (refresh) params.refresh = true;
    const response = await apiClient.get('/api/v1/signals/board', { params });
    const data = response.data as RawBoardResponse;
    return {
      asOf: data.as_of,
      entries: (data.entries ?? []).map(mapBoardEntry),
      counts: data.counts,
      degradedCodes: data.degraded_codes ?? [],
    };
  },

  /**
   * 拉取个股实时行情快照，映射 snake_case → camelCase。
   * code 经 encodeURIComponent 兼容带 '/' 的 crypto 代码（后端 {code:path} 路由），与 getSignals 同源。
   */
  async getQuote(code: string): Promise<StockQuote> {
    const response = await apiClient.get(`/api/v1/stocks/${encodeURIComponent(code)}/quote`);
    const d = response.data as {
      stock_code: string; stock_name: string | null; current_price: number;
      change: number | null; change_percent: number | null; open: number | null;
      high: number | null; low: number | null; prev_close: number | null;
      volume: number | null; amount: number | null; update_time: string | null;
    };
    return {
      stockCode: d.stock_code, stockName: d.stock_name ?? null, currentPrice: d.current_price,
      change: d.change ?? null, changePercent: d.change_percent ?? null, open: d.open ?? null,
      high: d.high ?? null, low: d.low ?? null, prevClose: d.prev_close ?? null,
      volume: d.volume ?? null, amount: d.amount ?? null, updateTime: d.update_time ?? null,
    };
  },
};

/** 后端 /history 返回的单根 K 线原始形状（snake_case，date 为 'YYYY-MM-DD'）。 */
export type KLineDataRaw = {
  date: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume?: number | null;
  amount?: number | null;
  change_percent?: number | null;
};

/**
 * 把后端 KLineData 映射为 klinecharts 所需的 KLine。
 *
 * - date 'YYYY-MM-DD'（无时区）统一锚 Asia/Shanghai 当日 00:00 → epoch ms，
 *   三市场（A股/港股/美股）一致，与 utils/format.ts 的 Asia/Shanghai 约定对齐；
 *   拼固定 +08:00 偏移而非依赖运行环境 TZ。
 * - amount → turnover；volume/amount 缺失归零（klinecharts 量副图需数值）。
 */
export const mapKLineDataToKLine = (raw: KLineDataRaw): KLine => ({
  timestamp: Date.parse(`${raw.date}T00:00:00+08:00`),
  open: raw.open,
  high: raw.high,
  low: raw.low,
  close: raw.close,
  volume: raw.volume ?? 0,
  turnover: raw.amount ?? 0,
});

File: api/v1/schemas/stocks.py
# -*- coding: utf-8 -*-
"""
===================================
股票数据相关模型
===================================

职责：
1. 定义股票实时行情模型
2. 定义历史 K 线数据模型
"""

from typing import Literal, List, Optional

from pydantic import BaseModel, ConfigDict, Field


class StockQuote(BaseModel):
    """股票实时行情"""
    
    stock_code: str = Field(..., description="股票代码")
    stock_name: Optional[str] = Field(None, description="股票名称")
    current_price: float = Field(..., description="当前价格")
    change: Optional[float] = Field(None, description="涨跌额")
    change_percent: Optional[float] = Field(None, description="涨跌幅 (%)")
    open: Optional[float] = Field(None, description="开盘价")
    high: Optional[float] = Field(None, description="最高价")
    low: Optional[float] = Field(None, description="最低价")
    prev_close: Optional[float] = Field(None, description="昨收价")
    volume: Optional[float] = Field(None, description="成交量（股）")
    amount: Optional[float] = Field(None, description="成交额（元）")
    update_time: Optional[str] = Field(None, description="更新时间")
    
    model_config = ConfigDict(json_schema_extra={
        "example": {
            "stock_code": "600519",
            "stock_name": "贵州茅台",
            "current_price": 1800.00,
            "change": 15.00,
            "change_percent": 0.84,
            "open": 1785.00,
            "high": 1810.00,
            "low": 1780.00,
            "prev_close": 1785.00,
            "volume": 10000000,
            "amount": 18000000000,
            "update_time": "2024-01-01T15:00:00"
        }
    })


class KLineData(BaseModel):
    """K 线数据点"""
    
    date: str = Field(..., description="日期")
    open: float = Field(..., description="开盘价")
    high: float = Field(..., description="最高价")
    low: float = Field(..., description="最低价")
    close: float = Field(..., description="收盘价")
    volume: Optional[float] = Field(None, description="成交量")
    amount: Optional[float] = Field(None, description="成交额")
    change_percent: Optional[float] = Field(None, description="涨跌幅 (%)")
    
    model_config = ConfigDict(json_schema_extra={
        "example": {
            "date": "2024-01-01",
            "open": 1785.00,
            "high": 1810.00,
            "low": 1780.00,
            "close": 1800.00,
            "volume": 10000000,
            "amount": 18000000000,
            "change_percent": 0.84
        }
    })


class ExtractItem(BaseModel):
    """单条提取结果（代码、名称、置信度）"""

    code: Optional[str] = Field(None, description="股票代码，None 表示解析失败")
    name: Optional[str] = Field(None, description="股票名称（如有）")
    confidence: str = Field("medium", description="置信度：high/medium/low")


class ExtractFromImageResponse(BaseModel):
    """图片股票代码提取响应"""

    codes: List[str] = Field(..., description="提取的股票代码（已去重，向后兼容）")
    items: List[ExtractItem] = Field(default_factory=list, description="提取结果明细（代码+名称+置信度）")
    raw_text: Optional[str] = Field(None, description="原始 LLM 响应（调试用）")


class StockHistoryResponse(BaseModel):
    """股票历史行情响应"""
    
    stock_code: str = Field(..., description="股票代码")
    stock_name: Optional[str] = Field(None, description="股票名称")
    period: str = Field(..., description="K 线周期")
    data: List[KLineData] = Field(default_factory=list, description="K 线数据列表")
    
    model_config = ConfigDict(json_schema_extra={
        "example": {
            "stock_code": "600519",
            "stock_name": "贵州茅台",
            "period": "daily",
            "data": []
        }
    })


class SignalMarker(BaseModel):
    """图上买卖信号标注点（追加契约，不影响旧 SniperPoints）"""

    timestamp: int = Field(..., description="epoch ms，唯一权威时间锚（前端按 timestamp 匹配蜡烛）")
    price: float = Field(..., description="标注价位")
    anchor: Literal["low", "high", "close"] = Field(..., description="price 锚定语义")
    direction: Literal["bullish", "bearish", "neutral"] = Field(..., description="信号方向")
    signal_type: str = Field(..., description="信号类型，如 volume_breakout / obv_top_divergence / llm_advice")
    source: Literal["rule", "llm"] = Field(..., description="信号来源")
    confidence: Literal["high", "medium", "low"] = Field(..., description="置信度（可排序，B 类<=low）")
    is_daily_approx: bool = Field(..., description="是否日线近似（VSA 等 B 类为 True）")
    is_anomalous: bool = Field(..., description="是否处于一字板/涨跌停等异常 bar")
    reason: str = Field(..., description="触发依据文案（用于钻取）")
    threshold: Optional[float] = Field(None, description="触发阈值，无则 null")
    observed_value: Optional[float] = Field(None, description="实际观测值，无则 null")
    hit_rate: Optional[float] = Field(None, description="历史方向命中率（M2c 回填），无样本则 null")
    hit_sample: Optional[int] = Field(None, description="命中率样本数（M2c 回填），无则 null")
    verified: bool = Field(False, description="hit_sample 达阈值且校正后下界 ci_low_corrected(family-wise Bonferroni;legacy 行回退 ci_low)> baseline 则 True(M3-A6/Inc 1c)")
    ci_low: Optional[float] = Field(None, description="命中率置信区间下界（M3-A6），无则 null")
    ci_high: Optional[float] = Field(None, description="命中率置信区间上界（M3-A6），无则 null")
    baseline_excess: Optional[float] = Field(None, description="相对基准超额（M3-A6），无则 null")
    ci_low_corrected: Optional[float] = Field(None, description="family-wise 多重检验校正后的 CI 下界(Inc 1c);null=legacy 行未重跑或无样本")
    family_size: Optional[int] = Field(None, description="该统计所在 family 的同检格子数 N(Inc 1c);null=legacy 行")
    as_of: Optional[int] = Field(None, description="仅 source=llm：该 LLM 结论生成时间 epoch ms")
    horizon_bars: Optional[int] = Field(None, description="该信号 hit_rate/CI 的验证前看窗口(bar 数);仅 rule、命中 signal_stats 时有值")
    status: Optional[Literal["active", "aging", "expired"]] = Field(None, description="信号生命周期;仅 rule。active=最新bar/aging=窗口内/expired=窗口已过")


class PriceLines(BaseModel):
    """买卖价位线（M2a 全 null，M2b 由价位反算器填值）"""

    entry: Optional[float] = Field(None, description="入场价位线，反算不出为 null")
    stop: Optional[float] = Field(None, description="止损价位线，反算不出为 null")
    target: Optional[float] = Field(None, description="目标价位线，反算不出为 null")


class SignalsResponse(BaseModel):
    """/signals 端点响应契约"""

    status: Literal["ok", "degraded"] = Field(..., description="整体状态；degraded 仍返回 200 + 部分结果")
    markers: List[SignalMarker] = Field(default_factory=list, description="信号标注点（rule 逐 bar + LLM 最新 1 点）")
    price_lines: PriceLines = Field(default_factory=PriceLines, description="买卖价位线（子字段允许 null）")
    consistency: Literal["consistent", "divergent", "conflict", "unknown", "stale"] = Field(
        ..., description="量价/规则与 LLM 一致性，仅在 LLM 点邻域计算"
    )
    degraded_reason: Optional[str] = Field(None, description="status=degraded 时的原因说明")
    resonance: Literal["none", "weekly", "weekly_monthly"] = Field(
        "none", description="多周期共振档位（M4-A）：高周期趋势与日线信号同向"
    )
    plan_quality: Optional[Literal["high", "medium", "low"]] = Field(None, description="交易计划质量:price_lines 完整度 + consistency;无 price_lines→null")

    model_config = ConfigDict(json_schema_extra={
        "example": {
            "status": "ok",
            "markers": [],
            "price_lines": {"entry": None, "stop": None, "target": None},
            "consistency": "consistent",
            "degraded_reason": None,
            "resonance": "none",
            "plan_quality": None,
        }
    })


class BoardEntry(BaseModel):
    """信号看板单行条目（容器 C / N1，追加契约）"""

    code: str = Field(..., description="股票代码")
    name: Optional[str] = Field(None, description="股票名称")
    market: Optional[str] = Field(None, description="市场，如 A/HK/US")
    action_group: Literal["buy", "hold", "sell", "unavailable"] = Field(
        ..., description="动作分组（看板分栏依据）"
    )
    rule_direction: Optional[Literal["bullish", "bearish", "neutral"]] = Field(
        None, description="规则/量价方向，无则 null"
    )
    llm_direction: Optional[Literal["bullish", "bearish", "neutral"]] = Field(
        None, description="LLM 方向，无则 null"
    )
    consistency: Literal["consistent", "divergent", "conflict", "unknown", "stale"] = Field(
        ..., description="规则与 LLM 一致性"
    )
    key_signals: List[str] = Field(default_factory=list, description="关键信号类型列表")
    price_lines: PriceLines = Field(..., description="买卖价位线（子字段允许 null）")
    latest_close: Optional[float] = Field(None, description="最新收盘价，无则 null")
    hit_rate: Optional[float] = Field(None, description="历史方向命中率，无样本则 null")
    hit_sample: Optional[int] = Field(None, description="命中率样本数，无则 null")
    verified: bool = Field(False, description="hit_sample 达阈值且校正后下界 ci_low_corrected(family-wise Bonferroni;legacy 行回退 ci_low)> baseline 则 True(M3-A6/Inc 1c)")
    ci_low: Optional[float] = Field(None, description="命中率置信区间下界（M3-A6），无则 null")
    ci_high: Optional[float] = Field(None, description="命中率置信区间上界（M3-A6），无则 null")
    baseline_excess: Optional[float] = Field(None, description="相对基准超额（M3-A6），无则 null")
    ci_low_corrected: Optional[float] = Field(None, description="family-wise 多重检验校正后的 CI 下界(Inc 1c);null=legacy 行未重跑或无样本")
    family_size: Optional[int] = Field(None, description="该统计所在 family 的同检格子数 N(Inc 1c);null=legacy 行")
    resonance: Literal["none", "weekly", "weekly_monthly"] = Field(
        "none", description="多周期共振档位（M4-A）"
    )
    status: Literal["ok", "degraded"] = Field(..., description="单条状态")
    degraded_reason: Optional[str] = Field(None, description="status=degraded 时的原因说明")
    horizon_bars: Optional[int] = Field(None, description="代表信号的验证前看窗口(bar 数),与本行 hit_rate 同源")
    signal_status: Optional[Literal["active", "aging", "expired"]] = Field(None, description="代表信号的生命周期(区别于 status 的 ok/degraded)")
    plan_quality: Optional[Literal["high", "medium", "low"]] = Field(None, description="交易计划质量(该股响应级)")


class BoardCounts(BaseModel):
    """看板各动作分组计数（容器 C / N1）"""

    buy: int = Field(0, description="buy 分组数量")
    hold: int = Field(0, description="hold 分组数量")
    sell: int = Field(0, description="sell 分组数量")
    unavailable: int = Field(0, description="unavailable 分组数量")


class SignalsBoardResponse(BaseModel):
    """信号看板端点响应契约（容器 C / N1）"""

    as_of: int = Field(..., description="看板数据基准时间 epoch ms")
    entries: List[BoardEntry] = Field(default_factory=list, description="看板条目列表")
    counts: BoardCounts = Field(..., description="各动作分组计数")
    degraded_codes: List[str] = Field(default_factory=list, description="降级的股票代码列表")
